daml-assistant/src/DAML/Assistant/Install/Github.hs

```diff
@@ -0,0 +1,112 @@
+-- Copyright (c) 2019 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+-- SPDX-License-Identifier: Apache-2.0
+
+{-# LANGUAGE OverloadedStrings #-}
+
+-- | Discover releases from the digital-asset/daml github.
+module DAML.Assistant.Install.Github
+    where
+
+import DAML.Assistant.Types
+import DAML.Assistant.Util
+import Data.Aeson
+import Network.HTTP.Simple
+import Control.Exception.Safe
+import Control.Monad
+import Data.List
+import Data.Either.Extra
+import qualified System.Info
+import qualified Data.Text as T
+
+newtype Tag = Tag Text deriving (Eq, Show, FromJSON)
+newtype AssetName = AssetName { unAssetName :: Text } deriving (Eq, Show, FromJSON)
+
+data Release = Release
+    { releaseTag          :: Tag
+    , releaseIsPrerelease :: Bool
+    , releaseAssets       :: [Asset]
+    } deriving (Eq, Show)
+
+instance FromJSON Release where
+    parseJSON = withObject "Release" $ \r ->
+        Release
+        <$> r .: "tag_name"
+        <*> r .: "prerelease"
+        <*> r .: "assets"
+
+data Asset = Asset
+    { assetName :: AssetName
+    , assetDownloadURL :: InstallURL
+    } deriving (Eq, Show)
+
+instance FromJSON Asset where
+    parseJSON = withObject "Asset" $ \r ->
+        Asset
+        <$> r .: "name"
+        <*> r .: "browser_download_url"
+
+versionToTag :: SdkVersion -> Tag
+versionToTag v = Tag ("v" <> versionToText v)
+
+tagToVersion :: Tag -> Either AssistantError SdkVersion
+tagToVersion (Tag t) =
+    mapLeft (assistantErrorBecause ("Tag " <> t <> "does not represent a valid SDK version.")) $
+        if T.take 1 t == "v" then
+            mapLeft (pack . displayException) $ parseVersion (T.drop 1 t)
+        else
+            Left "Tag must start with v followed by semantic version."
+
+
+makeAPIRequest :: FromJSON t => Text -> IO t
+makeAPIRequest path = do
+    request <- parseRequest (unpack ("GET https://api.github.com/repos/digital-asset/daml" <> path))
```

Comment:
GitHub has an API limit. How many simultaneous SDK users do we need to break it? Note that the HLint CI script has been known to break the GitHub API limit and thus start failing, so it isn't that high. If we're as popular as a relatively recent and less common feature of HLint we'll fail. I suspect this is an unworkable approach... (as in you'd have to mirror it, cache it, create a github token or something)

Reply:
Repeating what we discussed on slack: Github has a limit of 60 requests per hour per IP for unauthenticated requests. An individual user is unlikely to hit this but we risk hitting this when IPs are shared, e.g., on Travis or if lots of people sit behind the same NAT. For now, it is probably fine. It should be easy to setup a mirror if we run into this.